crates/wash-lib/src/parser/mod.rs

```diff
@@ -138,13 +140,222 @@ pub struct RustConfig {
+    /// or special-cased in some other way
+    /// (e.g. `wasi:http` is a well known standard)
+    #[default]
+    Builtin,
+
+    /// A file source
+    ///
+    /// These references are resolved in two ways:
+    ///   - If a directory, then the namespace & path are appended
+    ///   - If a direct file then the file itself is used
+    ///
+    /// (ex. 'file://relative/path/to/file', 'file:///absolute/path/to/file')
+    LocalPath(String),
+
+    /// Remote HTTP registry, configured to support `.well-known/wasm-pkg/registry.json`
+    RemoteHttpWellKnown(String),
+
+    /// An OCI reference
+    ///
+    /// These references are resolved by appending the intended namespace and package
+    /// to the provided URI
+    ///
+    /// (ex. resolving `wasi:keyvalue@0.2.0` with 'oci://ghcr.io/wasmcloud/wit' becomes `oci://ghcr.io/wasmcloud/wit/wasi/keyvalue:0.2.0`)
+    RemoteOci(String),
+
+    /// URL to a HTTP/S resource
+    ///
+    /// These references are resolved by downloading and uncompressing (where possible) the linked file
+    /// as WIT, for whatever interfaces were provided.
+    ///
+    /// (ex. resolving `https://example.com/wit/package.tgz` means downloading and unpacking the tarball)
+    RemoteHttp(String),
+
+    /// URL to a GIT repository
+    ///
+    /// These URLs are guaranteed to start with a git-related scheme (ex. `git+http://`, `git+ssh://`, ...)
+    /// and will be used as the base under which to pull a folder of WIT
+    RemoteGit(String),
+}
+
+impl Display for RegistryPullSource {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            RegistryPullSource::Builtin => write!(f, "builtin")?,
+            RegistryPullSource::LocalPath(s)
+            | RegistryPullSource::RemoteHttpWellKnown(s)
+            | RegistryPullSource::RemoteOci(s)
+            | RegistryPullSource::RemoteHttp(s)
+            | RegistryPullSource::RemoteGit(s) => write!(f, "{}", s)?,
+        }
+        Ok(())
+    }
+}
+
+impl<'de> Deserialize<'de> for RegistryPullSource {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Self::try_from(String::deserialize(deserializer)?).map_err(serde::de::Error::custom)
+    }
+}
+
+impl TryFrom<String> for RegistryPullSource {
+    type Error = anyhow::Error;
+
+    fn try_from(value: String) -> Result<Self> {
+        Self::from_str(&value)
+    }
+}
+
+impl FromStr for RegistryPullSource {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        Ok(match s {
+            s if s.starts_with("file://") => Self::LocalPath(s.into()),
+            s if s.starts_with("oci://") => Self::RemoteOci(s.into()),
+            s if s.starts_with("http://") || s.starts_with("https://") => {
+                Self::RemoteHttp(s.into())
+            }
+            s if s.starts_with("git+ssh://")
```

Comment:
In the other code I saw the shortcut with leaving off `github.com`. Does that need to be handled here or is it really only for `wash new`?

Reply:
Nope, I don't think we should allow leaving off github.com here -- people should be explicit about the git repository they're accessing.